OpenRA.Mods.Cnc/Activities/LayMines.cs

```diff
@@ -32,21 +32,42 @@ public class LayMines : Activity
 		bool returnToBase;
 		Actor rearmTarget;
 
-		public LayMines(Actor self, CPos[] minefield)
+		public LayMines(Actor self, List<CPos> minefield = null)
 		{
 			minelayer = self.Trait<Minelayer>();
 			ammoPools = self.TraitsImplementing<AmmoPool>().ToArray();
 			movement = self.Trait<IMove>();
 			rearmableInfo = self.Info.TraitInfoOrDefault<RearmableInfo>();
-
-			if (minefield != null)
-				this.minefield = minefield.ToList();
+			this.minefield = minefield;
 		}
 
 		protected override void OnFirstRun(Actor self)
 		{
 			if (minefield == null)
-				minefield = new CPos[] { self.Location }.ToList();
+				minefield = new List<CPos> { self.Location };
+		}
+
+		CPos? NextValidCell(Actor self)
+		{
+			if (minefield == null)
+				return null;
+
+			// Assume that the closest cell will be valid
+			var candidateCells = minefield;
+			while (candidateCells.Count > 0)
```

Comment:
Would it be cleaner to use Queue<> here?

Reply:
Probably not - we aren't guaranteed to access the cells in order, because the earlier cells in the list may be transiently blocked. The slightly weird syntax here is designed to avoid copying in the common case where it isn't blocked, so turning it to a queue here would ruin that.

Reply:
This whole thing *can* probably simplify down to a LINQ statement now that we aren't manipulating anything, though.